critiquebrainz/db/statistics.py

```diff
@@ -0,0 +1,255 @@
+        commenters = [dict(commenter) for commenter in commenters]
+        return commenters
+
+
+def get_top_users(from_date=date(1970, 1, 1), to_date=date.today()+timedelta(1), review_weight=1,
+                  comment_weight=1, vote_weight=1, limit=10):
+    """ Gets list of top contributors based on number of reviews, votes and comments
+        along with their final scores.
+        score = (reviews * review_weight + comments * comment_weight + votes * vote_weight)
+        Results are sorted in ascending order by score and max number of results are
+        defined by 'limit'.
+
+    Args:
+        from_date(datetime): Date from which contributions by users are to be considered.
+        to_date(datetime): Date upto which contributions by users are to be considered.
+        review_weight(int): Weight for each review of a user to add to their final score
+        comment_weight(int): Weight for each comment of a user to add to their final score
+        vote_weight(int): Weight for each vote of a user to add to their final score
+
+    Returns:
+        List of dictionaries where each dictionary has the following structure:
+        {
+            "id": (uuid),
+            "display_name": (str),
+            "review_count": (int),
+            "comment_count": (int),
+            "vote_count": (int),
+            "score": (int),
+        }
+    """
+
+    reviewers = get_users_with_review_count(from_date=from_date, to_date=to_date)
+    commenters = get_users_with_comment_count(from_date=from_date, to_date=to_date)
+    voters = get_users_with_vote_count(from_date=from_date, to_date=to_date)
+
+    # merge based on user_id
+    top_scorers = merge_rows(merge_rows(reviewers, commenters, "id"), voters, "id")
+
+    # add 'score' for each user
+    for user in top_scorers:
+        user["id"] = str(user["id"])
+        user["score"] = user["review_count"]*review_weight + user["comment_count"]*comment_weight + user["vote_count"]*vote_weight
+
+    # sort top_users by 'score' in descending order and keep only top 'limit' users
+    top_scorers = sorted(top_scorers, key=lambda row: row["score"], reverse=True)[:limit]
+    if top_scorers[0]["score"] == 0:
+        top_scorers = []
+    return top_scorers
+
+
+def get_top_users_overall():
+    """ Gets top contributors since the beginning
+
+    Returns:
+        Returns:
+        List of dictionaries where each dictionary has the following structure:
+        {
+            "id": (str),
+            "display_name": (str),
+            "review_count": (int),
+            "comment_count": (int),
+            "vote_count": (int),
+            "score": (int),
+        }
+    """
+    key = cache.gen_key("top_users_overall", _CACHE_NAMESPACE)
+    top_users = cache.get(key, _CACHE_NAMESPACE)
+
+    # if could not fetch results from cache, or fetched results have to be updated
+    if not top_users:
+
+        try:
+            results = get_top_users(
+                review_weight=5,
```

Comment:
magic numbers are generally not a good idea, these should be configurable. Let's put them at the top of this file as constants.

Reply:
done!